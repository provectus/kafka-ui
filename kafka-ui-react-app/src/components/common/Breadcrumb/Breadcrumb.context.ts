import { createContext } from 'react';

export interface BreadcrumbEntry {
  link: string;
  path: string[];
}

interface BreadcrumbContextInterface extends BreadcrumbEntry {
  handleRouteChange: (match: { url: string; path: string }) => void;
}

export const BreadcrumbContext = createContext<BreadcrumbContextInterface>({
  link: '',
  path: [],
  handleRouteChange: () => {},
});
