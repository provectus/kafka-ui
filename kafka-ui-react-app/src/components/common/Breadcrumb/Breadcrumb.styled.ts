import styled from 'styled-components';
import { Colors } from 'theme/theme';

export const BreadcrumbWrapper = styled.ul`
  display: flex;
  padding-left: 16px;
  padding-top: 1em;

  font-size: 12px;

  & li:not(:last-child)::after {
    content: '/';
    color: ${Colors.neutral[30]};
    margin: 0 8px;
  }
`;
