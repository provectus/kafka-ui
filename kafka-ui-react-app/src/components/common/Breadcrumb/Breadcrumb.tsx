import React from 'react';
import { NavLink } from 'react-router-dom';

export interface Link {
  label: string;
  href: string;
}

interface Props {
  links?: Link[];
}

const Breadcrumb: React.FC<Props> = ({ links, children }) => {
  return (
    <nav className="breadcrumb" aria-label="breadcrumbs">
      <ul>
        {links &&
          links.map(({ label, href }) => (
            <li key={href}>
              <NavLink to={href}>{label}</NavLink>
            </li>
          ))}

        <li className="is-active">
          <span className="">{children}</span>
        </li>
      </ul>
    </nav>
  );
};

export default Breadcrumb;
