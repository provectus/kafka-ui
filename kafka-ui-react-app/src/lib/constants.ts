import { SelectOption } from 'components/common/Select/Select';
import { ConfigurationParameters } from 'generated-sources';

declare global {
  interface Window {
    basePath: string;
  }
}

export const BASE_PARAMS: ConfigurationParameters = {
  basePath: window.basePath || '',
  credentials: 'include',
  headers: {
    'Content-Type': 'application/json',
  },
};

export const TOPIC_NAME_VALIDATION_PATTERN = /^[a-zA-Z0-9._-]+$/;
export const SCHEMA_NAME_VALIDATION_PATTERN = /^[.,A-Za-z0-9_/-]+$/;

export const TOPIC_CUSTOM_PARAMS_PREFIX = 'customParams';
export const TOPIC_CUSTOM_PARAMS: Record<string, string> = {
  'compression.type': 'producer',
  'leader.replication.throttled.replicas': '',
  'message.downconversion.enable': 'true',
  'segment.jitter.ms': '0',
  'flush.ms': '9223372036854775807',
  'follower.replication.throttled.replicas': '',
  'segment.bytes': '1073741824',
  'flush.messages': '9223372036854775807',
  'message.format.version': '2.3-IV1',
  'file.delete.delay.ms': '60000',
  'max.compaction.lag.ms': '9223372036854775807',
  'min.compaction.lag.ms': '0',
  'message.timestamp.type': 'CreateTime',
  preallocate: 'false',
  'min.cleanable.dirty.ratio': '0.5',
  'index.interval.bytes': '4096',
  'unclean.leader.election.enable': 'true',
  'retention.bytes': '-1',
  'delete.retention.ms': '86400000',
  'segment.ms': '604800000',
  'message.timestamp.difference.max.ms': '9223372036854775807',
  'segment.index.bytes': '10485760',
};

export const MILLISECONDS_IN_WEEK = 604_800_000;
export const MILLISECONDS_IN_DAY = 86_400_000;
export const MILLISECONDS_IN_SECOND = 1_000;

export const NOT_SET = -1;
export const BYTES_IN_GB = 1_073_741_824;
export const BUILD_VERSION_PATTERN = /v\d.\d.\d/;

export const PER_PAGE = 25;
export const MESSAGES_PER_PAGE = '100';

export const GIT_REPO_LINK = 'https://github.com/provectus/kafka-ui';
export const GIT_REPO_LATEST_RELEASE_LINK =
  'https://api.github.com/repos/provectus/kafka-ui/releases/latest';

export const LOCAL_STORAGE_KEY_PREFIX = 'kafka-ui';

export enum AsyncRequestStatus {
  initial = 'initial',
  pending = 'pending',
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export const QUERY_REFETCH_OFF_OPTIONS = {
  refetchOnMount: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
};

// Cluster Form Constants
export const AUTH_OPTIONS: SelectOption[] = [
  { value: 'SASL/JAAS', label: 'SASL/JAAS' },
  { value: 'SASL/GSSAPI', label: 'SASL/GSSAPI' },
  { value: 'SASL/OAUTHBEARER', label: 'SASL/OAUTHBEARER' },
  { value: 'SASL/PLAIN', label: 'SASL/PLAIN' },
  { value: 'SASL/SCRAM-256', label: 'SASL/SCRAM-256' },
  { value: 'SASL/SCRAM-512', label: 'SASL/SCRAM-512' },
  { value: 'Delegation tokens', label: 'Delegation tokens' },
  { value: 'SASL/LDAP', label: 'SASL/LDAP' },
  { value: 'SASL/AWS IAM', label: 'SASL/AWS IAM' },
  { value: 'mTLS', label: 'mTLS' },
];

export const SECURITY_PROTOCOL_OPTIONS: SelectOption[] = [
  { value: 'SASL_SSL', label: 'SASL_SSL' },
  { value: 'SASL_PLAINTEXT', label: 'SASL_PLAINTEXT' },
];
export const METRICS_OPTIONS: SelectOption[] = [
  { value: 'JMX', label: 'JMX' },
  { value: 'PROMETHEUS', label: 'PROMETHEUS' },
];
