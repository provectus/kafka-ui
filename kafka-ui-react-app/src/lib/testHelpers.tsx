import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
