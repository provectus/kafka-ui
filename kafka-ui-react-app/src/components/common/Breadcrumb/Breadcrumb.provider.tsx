import React, { useState } from 'react';
import capitalize from 'lodash/capitalize';

import { BreadcrumbContext, BreadcrumbEntry } from './Breadcrumb.context';

const mapLocationToPath = (
  splittedLocation: string[],
  splittedRoutePath: string[]
) =>
  splittedLocation.map((item, index) =>
    splittedRoutePath[index]?.charAt(0) !== ':'
      ? item.split('-').map(capitalize).join(' ')
      : item
  );

export const BreadcrumbProvider: React.FC = ({ children }) => {
  const [state, setState] = useState<BreadcrumbEntry>({
    link: '',
    path: [],
  });

  const handleRouteChange = (params: { url: string; path: string }) => {
    setState((prevState) => {
      const newState = { ...prevState };
      const splittedRoutePath = params.path.split('/');
      const splittedLocation = params.url.split('/');

      if (prevState.link !== params.url) {
        newState.link = params.url;
        newState.path = mapLocationToPath(splittedLocation, splittedRoutePath);
      }

      if (prevState.path.length < params.path.split('/').length) {
        newState.path = mapLocationToPath(splittedLocation, splittedRoutePath);
      }

      return newState;
    });
  };

  return (
    <BreadcrumbContext.Provider
      value={{
        ...state,
        handleRouteChange,
      }}
    >
      {children}
    </BreadcrumbContext.Provider>
  );
};
