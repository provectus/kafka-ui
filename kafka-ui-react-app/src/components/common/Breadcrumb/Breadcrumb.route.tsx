import React, { useContext, useEffect } from 'react';
import {
  Route,
  RouteProps,
  useLocation,
  useRouteMatch,
} from 'react-router-dom';

import { BreadcrumbContext } from './Breadcrumb.context';

const BreadcrumbRouteInternal: React.FC = () => {
  const match = useRouteMatch();
  const location = useLocation();
  const context = useContext(BreadcrumbContext);

  useEffect(() => {
    context.handleRouteChange({ ...match, url: location.pathname });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [location.pathname]);

  return null;
};

export const BreadcrumbRoute: React.FC<RouteProps> = ({
  children,
  render,
  component,
  ...props
}) => {
  return (
    <Route
      {...props}
      render={(routeParams) => {
        if (component) {
          return (
            <>
              {React.createElement(component)}
              <BreadcrumbRouteInternal />
            </>
          );
        }
        if (render) {
          return (
            <>
              {render(routeParams)}
              <BreadcrumbRouteInternal />
            </>
          );
        }

        return (
          <>
            {children}
            <BreadcrumbRouteInternal />
          </>
        );
      }}
    />
  );
};
