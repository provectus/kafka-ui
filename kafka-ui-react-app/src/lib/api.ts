import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
