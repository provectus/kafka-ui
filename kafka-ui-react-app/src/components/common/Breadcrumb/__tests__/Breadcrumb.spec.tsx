import React from 'react';
import Breadcrumb from 'components/common/Breadcrumb/Breadcrumb';
import { BreadcrumbProvider } from 'components/common/Breadcrumb/Breadcrumb.provider';
import { BreadcrumbRoute } from 'components/common/Breadcrumb/Breadcrumb.route';
import { render } from 'lib/testHelpers';

const createTopicPath = '/clusters/local/topics/create-new';
const createTopicRoutePath = '/clusters/:clusterName/topics/create-new';

const topicPath = '/clusters/secondLocal/topics/topic-name';
const topicRoutePath = '/clusters/:clusterName/topics/:topicName';

describe('Breadcrumb component', () => {
  const setupComponent = (pathname: string, routePath: string) =>
    render(
      <BreadcrumbProvider>
        <Breadcrumb />
        <BreadcrumbRoute path={routePath} />
      </BreadcrumbProvider>,
      { pathname }
    );

  it('renders the list of links', async () => {
    const { getByText } = setupComponent(createTopicPath, createTopicRoutePath);
    expect(getByText('Topics')).toBeInTheDocument();
    expect(getByText('Create New')).toBeInTheDocument();
  });
  it('renders the topic overview', async () => {
    const { getByText } = setupComponent(topicPath, topicRoutePath);
    expect(getByText('Topics')).toBeInTheDocument();
    expect(getByText('topic-name')).toBeInTheDocument();
  });
});
