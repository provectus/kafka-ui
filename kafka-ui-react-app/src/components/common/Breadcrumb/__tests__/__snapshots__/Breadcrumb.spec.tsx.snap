// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb component matches the snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <Breadcrumb
    links={
      Array [
        Object {
          "href": "link1href",
          "label": "link1",
        },
        Object {
          "href": "link2href",
          "label": "link2",
        },
        Object {
          "href": "link3href",
          "label": "link3",
        },
      ]
    }
  >
    <div
      className="child"
    />
  </Breadcrumb>
</Router>
`;
